tests/unit/onyxClearNativeStorageTest.js

```diff
@@ -0,0 +1,162 @@
+    DEFAULT_KEY: 'defaultKey',
+};
+const SET_VALUE = 'set';
+const MERGED_VALUE = 'merged';
+const DEFAULT_VALUE = 'default';
+
+let storageCallResolveList = [];
+
+/**
+ * Get the order in which the storage method call resolved so we can assert that
+ * updates to storage are made in the expected order.
+ * @param {String} methodName The name of the storage method
+ * @returns {Number}
+ */
+function getStorageCallResolveOrder(methodName) {
+    return storageCallResolveList.indexOf(methodName) + 1;
+}
+
+// Add all storage calls to a queue and make them wait for each other to match the real implementation.
+let storageCallQueue = [];
+
+// Track when AsyncStorageMock.clear calls resolve.
+const asyncStorageMockClear = AsyncStorageMock.clear;
+AsyncStorageMock.clear = jest.fn(() => Promise.all(storageCallQueue)
+    .then(() => {
+        const clearPromise = asyncStorageMockClear()
+            .then(storageCallResolveList.push('clear'));
+        storageCallQueue.push(clearPromise);
+        return clearPromise;
+    }));
+
+// Track when AsyncStorageMock.setItem calls resolve.
+const asyncStorageMockSetItem = AsyncStorageMock.setItem;
+AsyncStorageMock.setItem = jest.fn((key, value) => Promise.all(storageCallQueue)
+    .then(() => {
+        const setItemPromise = asyncStorageMockSetItem(key, value)
+            .then(storageCallResolveList.push('setItem'));
+        storageCallQueue.push(setItemPromise);
+        return setItemPromise;
+    }));
+
+describe('Set data while storage is clearing', () => {
+    let connectionID;
+    let onyxValue;
+
+    /** @type OnyxCache */
+    let cache;
+
+    // Always use a "fresh" cache instance
+    beforeEach(() => {
+        onyxValue = null;
+        cache = require('../../lib/OnyxCache').default;
+        Onyx.init({
+            keys: ONYX_KEYS,
+            registerStorageEventListener: () => {},
+            initialKeyStates: {
+                [ONYX_KEYS.DEFAULT_KEY]: DEFAULT_VALUE,
+            },
+        });
+        connectionID = Onyx.connect({
+            key: ONYX_KEYS.DEFAULT_KEY,
+            initWithStoredValues: false,
+            callback: val => onyxValue = val,
+        });
+        return waitForPromisesToResolve();
+    });
+
+    afterEach(() => {
+        Onyx.disconnect(connectionID);
+        Storage.clear.mockReset();
+        return Onyx.clear()
+            .then(() => {
+                storageCallResolveList = [];
+                storageCallQueue = [];
+            });
+    });
+
+    it('should persist the value of Onyx.merge when called between the cache and storage clearing', () => {
+        expect.assertions(5);
+
+        // GIVEN that Onyx is completely clear
+        // WHEN Onyx.clear() is called
+        Storage.clear = jest.fn(() => {
+            // WHEN merge is called between the cache and storage clearing, on a key with a default key state
+            Onyx.merge(ONYX_KEYS.DEFAULT_KEY, MERGED_VALUE);
+            AsyncStorageMock.clear();
```

Comment:
I'm not following how this is possible or why we are creating a mock of `Storage.clear()` that does a weird thing that never actually happens. So I think we should test the actual code that will run and simplify this test.

The examples I had in mind for something like this were more like:

```js
const clearingPromise = Onyx.clear();
Onyx.merge('test', {test: 1});
return clearingPromise.then(() => {
    expect(onyxValue).toEqual({test: 1});
});
``` 
Tthat should replicate the condition where `Onyx.clear()` was called and then `Onxy.merge()` was called _before_ the `Storage` was actually cleared.

Otherwise, I'm not sure what situation we're trying to test exactly. I'm making some assumptions (i.e. that we don't have to dip into storage to get any values) but this is what I think the order of operations would typically be...

- `clear()` called and `getAllKeys()` is called (returns a promise so the merge logic is what runs next)
- `merge()` called and we call `get()` here (also returns a promise)
https://github.com/Expensify/react-native-onyx/blob/f11e7c1949916a81c36c17dd87919dd532d4620f/lib/Onyx.js#L672
- I would think that call to get will pull the cached value _before_  any of these keys are modified here
https://github.com/Expensify/react-native-onyx/blob/a37d7661a8702d75eb6244fb6ff6f47d004e1eb6/lib/Onyx.js#L721-L723
- The `getAllKeys()` promise will returns next and the above logic runs ([we shouldn't have to dip into storage to get these keys so it will just return](https://github.com/Expensify/react-native-onyx/blob/a37d7661a8702d75eb6244fb6ff6f47d004e1eb6/lib/Onyx.js#L75-L78))
- `Storage.clear()` promise starts
- Then the `get()` call will return to trigger the merge and call `setItem()` [here](https://github.com/Expensify/react-native-onyx/blob/a37d7661a8702d75eb6244fb6ff6f47d004e1eb6/lib/Onyx.js#L682) this will modify the values in the cache and start the `Storage.setItem()` promise.
- At this point, the cache should be clear except for the merged value?

Reply:
> I'm not following how this is possible or why we are creating a mock of Storage.clear() that does a weird thing that never actually happens. So I think we should test the actual code that will run and simplify this test.

I was doing this to make sure that `merge()` is called between the cache and storage clearing. Your explanation of the flow makes sense, so I guess I don't need that mock.